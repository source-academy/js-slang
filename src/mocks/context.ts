import * as es from 'estree'

import createContext from '../createContext'
import Closure from '../interpreter/closure'
import { Context, Environment, Variant } from '../types'

export function mockContext(chapter = 1, variant: Variant = 'default'): Context {
  return createContext(chapter, variant)
}

export function mockRuntimeContext(): Context {
  const context = createContext()
  context.runtime = {
    break: false,
    debuggerOn: true,
    isRunning: true,
    environments: [],
    nodes: [
      {
        type: 'Literal',
        loc: {
          start: { line: 1, column: 0 },
          end: { line: 1, column: 1 }
        },
        value: 0,
        raw: '0',
        range: [0, 1]
      }
    ]
  }
  return context
}

export function mockClosure(): Closure {
  return new Closure(
    {
      type: 'FunctionExpression',
      loc: null,
      id: null,
      params: [],
      body: {
        type: 'BlockStatement',
        body: []
      }
    } as es.FunctionExpression,
    {} as Environment,
    {} as Context
  )
}
