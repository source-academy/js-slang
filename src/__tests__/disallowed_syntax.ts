import { mockContext } from '../mocks/context'
import { parseError, runInContext } from '../index'

test("Cannot leave blank expressions in for loops", () => {
  const code = [
    `
    for(; i < 3; i = i + 1) {
      break;
    }
  `,
    `
    for(let i = 0; ; i = i + 1) {
      break;
    }
  `,
    `
    for(let i = 0; i < 3;) {
      break;
    }
  `,
    `
    for(;;) {
      break;
    }
  `
  ];
  const scheduler = "preemptive";
  const promises = code.map(c => {
    const context = mockContext(3);
    return runInContext(c, context, { scheduler }).then(obj => ({
      context,
      obj
    }));
  });
  return Promise.all(promises).then(results => {
    results.map(res => {
      const { context, obj } = res;
      expect(obj.status).toBe("error");
      const errors = parseError(context.errors);
      expect(errors).toMatchSnapshot();
    });
  });
});

test("Cannot use update expressions", () => {
  const code = `
  let x = 3;
  x++;
  x;
  `;
  const context = mockContext(3);
  const promise = runInContext(code, context, { scheduler: "preemptive" });
  return promise.then(obj => {
    expect(obj.status).toBe("error");
    const errors = parseError(context.errors);
    expect(errors).toMatchSnapshot();
  });
});

