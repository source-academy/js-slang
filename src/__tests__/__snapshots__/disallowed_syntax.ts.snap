// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cannot have if without else 1`] = `"Line 2: Missing \\"else\\" in \\"if-else\\" statement"`;

exports[`Cannot have incomplete statements 1`] = `"Line 2: Missing semicolon at the end of statement"`;

exports[`Cannot leave blank expressions in for loops 1`] = `"Line 2: Missing init expression in for statement."`;

exports[`Cannot leave blank expressions in for loops 2`] = `"Line 2: Missing test expression in for statement."`;

exports[`Cannot leave blank expressions in for loops 3`] = `"Line 2: Missing update expression in for statement."`;

exports[`Cannot leave blank expressions in for loops 4`] = `"Line 2: Missing init, test, update expressions in for statement."`;

exports[`Cannot leave while loop predicate blank 1`] = `"Line 2: SyntaxError: Unexpected token (2:8)"`;

exports[`Cannot use assignment expressions 1`] = `"Line 3: Assignment inside an expression is not allowed. Only assignment in a statement is allowed."`;

exports[`Cannot use assignment expressions 2`] = `"Line 4: Assignment inside an expression is not allowed. Only assignment in a statement is allowed."`;

exports[`Cannot use assignment expressions 3`] = `"Line 3: Assignment inside an expression is not allowed. Only assignment in a statement is allowed."`;

exports[`Cannot use destructuring declarations 1`] = `"Line 3: Array patterns are not allowed"`;

exports[`Cannot use function expressions 1`] = `"Line 2: Function expressions are not allowed"`;

exports[`Cannot use function expressions 2`] = `"Line 2: Function expressions are not allowed"`;

exports[`Cannot use multiple declarations 1`] = `"Line 2: Multiple declaration in a single statement"`;

exports[`Cannot use update expressions 1`] = `""`;

exports[`Cannot use update statements 1`] = `"Line 3: The assignment operator += is not allowed. Use = instead"`;

exports[`No array expressions in chapter 2 1`] = `"Line 2: Array expressions are not allowed"`;

exports[`No empty statements 1`] = `"Line 2: Empty statements are not allowed"`;

exports[`No rest pattern 1`] = `"Line 2: Rest elements are not allowed"`;

exports[`No spread operator 1`] = `"Line 5: Spread elements are not allowed"`;

exports[`No trailing commas in arrays 1`] = `
"Line 2: Trailing comma
Line 2: Array expressions are not allowed"
`;

exports[`No trailing commas in objects 1`] = `"Line 4: SyntaxError: Unexpected token (4:7)"`;

exports[`for needs braces 1`] = `
Object {
  "status": "error",
}
`;

exports[`for needs braces 2`] = `"Line 2: Missing curly braces around \\"for\\" block"`;

exports[`if needs braces 1`] = `
Object {
  "status": "error",
}
`;

exports[`if needs braces 2`] = `
"Line 2: Missing curly braces around \\"if\\" block
Line 2: Missing curly braces around \\"else\\" block"
`;

exports[`no assigning to reserved keywords 1`] = `"Line 2: SyntaxError: The keyword 'package' is reserved (2:4)"`;

exports[`no classes 1`] = `
"Line 2: Class bodys are not allowed
Line 2: Class declarations are not allowed"
`;

exports[`no debugger statement 1`] = `"Line 2: Debugger statements are not allowed"`;

exports[`no declaration without assignment 1`] = `"Line 2: Missing value in variable declaration"`;

exports[`no declaring reserved keywords 1`] = `"Line 2: SyntaxError: The keyword 'yield' is reserved (2:8)"`;

exports[`no export constant 1`] = `"Line 2: Export named declarations are not allowed"`;

exports[`no export default 1`] = `"Line 3: Export default declarations are not allowed"`;

exports[`no export function 1`] = `"Line 2: Export named declarations are not allowed"`;

exports[`no for in loops 1`] = `
"Line 2: Missing value in variable declaration
Line 2: For in statements are not allowed"
`;

exports[`no for of loops 1`] = `
"Line 2: Missing value in variable declaration
Line 2: For of statements are not allowed"
`;

exports[`no generator functions 1`] = `"Line 3: Yield expressions are not allowed"`;

exports[`no implicit undefined return 1`] = `"Line 3: Missing value in return statement"`;

exports[`no import 1`] = `
"Line 2: Import specifiers are not allowed
Line 2: Import declarations are not allowed"
`;

exports[`no interface 1`] = `"Line 2: SyntaxError: The keyword 'interface' is reserved (2:4)"`;

exports[`no regexp 1`] = `
"Line 1: Missing semicolon at the end of statement
Line 1: 'RegExp' literals are not allowed"
`;

exports[`no repeated params 1`] = `"Line 2: SyntaxError: Argument name clash (2:18)"`;

exports[`no sequence expression 1`] = `"Line 2: Sequence expressions are not allowed"`;

exports[`no super 1`] = `
"Line 4: Supers are not allowed
Line 3: Function expressions are not allowed
Line 3: Method definitions are not allowed
Line 2: Class bodys are not allowed
Line 2: Class declarations are not allowed"
`;

exports[`no template literals 1`] = `
"Line 1: Missing semicolon at the end of statement
Line 1: Template elements are not allowed
Line 1: Template literals are not allowed"
`;

exports[`no this, no new 1`] = `"Line 3: 'this' expressions are not allowed"`;

exports[`no try statements 1`] = `
"Line 6: Spread elements are not allowed
Line 7: Catch clauses are not allowed
Line 5: Try statements are not allowed"
`;

exports[`no unspecified operators 1`] = `"Line 2: Operator '<<' is not allowed."`;

exports[`no unspecified unary operators 1`] = `"Line 3: Operator 'typeof' is not allowed."`;

exports[`while needs braces 1`] = `
Object {
  "status": "error",
}
`;

exports[`while needs braces 2`] = `"Line 3: Missing curly braces around \\"while\\" block"`;
